sync/src/snapshot/service.rs

```diff
@@ -66,11 +73,26 @@ impl ChainNotify for Service {
             let path: PathBuf = [self.root_dir.clone(), format!("{:x}", *header.hash())].iter().collect();
             let root = header.state_root();
             // FIXME: The db can be corrupted because the CodeChain doesn't wait child threads end on exit.
-            spawn(move || match Snapshot::try_new(path).map(|s| s.write_snapshot(db.as_ref(), &root)) {
-                Ok(_) => {}
-                Err(Error::FileError(ErrorKind::AlreadyExists)) => {}
-                Err(e) => cerror!(SNAPSHOT, "{}", e),
+            let id = self.thread_ids.fetch_add(1, Ordering::SeqCst);
+            let joins = Arc::clone(&self.joins);
+            let join = spawn(move || {
+                match Snapshot::try_new(path).map(|s| s.write_snapshot(db.as_ref(), &root)) {
+                    Ok(_) => {}
+                    Err(Error::FileError(ErrorKind::AlreadyExists)) => {}
+                    Err(e) => cerror!(SNAPSHOT, "{}", e),
+                }
+                joins.lock().remove(&id);
             });
+            self.joins.lock().insert(id, join);
```

Comment:
There might be a race condition where the spawned thread exits first, so `joins.remove()` is called first and `joins.insert()` is called later.

How about cleaning it lazily?
```rust
fn new_blocks() {
	...
	let arc = Arc::new(());
	let weak = Arc::downgrade(&arc);
	let join = spawn(move || {
		let arc = arc; // If the stack dies, the arc dies. 
	    ...
	}
	let mut joins = joins.lock();
	joins.push((weak, join));
	joins.retain(|(weak, _)| Weak::upgrade(weak).is_some());
	...
}
```

Reply:
I think it can be solved by locking `self.joins` in the upper line.
```rust
let joins = Arc::clone(&self.joins);
let self_join = self.joins.lock();
let join = spawn(move || { ... });
self_join.insert(id, join);
```
The variable name is a bit awkward, but I think you'll get the point.

Reply:
@foriequal0 @Remagpie How about it?
```rust
let id = self.thread_ids.fetch_add(1, Ordering::SeqCst);
self.joins.lock().insert(id, join);
let joins = Arc::clone(&self.joins);
let join = spawn(move || {
```
I think it's simple but working.

Reply:
IMO, @Remagpie's looks simple. @foriequal0 how about it?